File: web/lib/measurement/tracking.ts
const encoder = new TextEncoder();

export async function hashTelegramId(
  telegramId: string | number,
  secret: string,
): Promise<string> {
  if (!secret) throw new Error("USER_HASH_SECRET не задан");
  const key = await crypto.subtle.importKey(
    "raw",
    encoder.encode(secret),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"],
  );
  const signature = await crypto.subtle.sign(
    "HMAC",
    key,
    encoder.encode(String(telegramId)),
  );
  return Array.from(new Uint8Array(signature))
    .map((byte) => byte.toString(16).padStart(2, "0"))
    .join("")
    .slice(0, 32);
}

export function buildDeepLink(botUsername: string, token: string): string {
  const username = botUsername.trim().replace(/^@/, "");
  if (!username) throw new Error("BOT_USERNAME не задан");
  if (!token.trim()) throw new Error("Tracking token не задан");
  return `https://t.me/${username}?start=${encodeURIComponent(token.trim())}`;
}

export function newTrackingToken(): string {
  return `p_${crypto.randomUUID().replaceAll("-", "").slice(0, 12)}`;
}

File: src/tracking.py
import hashlib
import hmac
import os
import secrets

def _hash_secret() -> bytes:
    secret = os.environ.get("USER_HASH_SECRET", "").strip()
    if not secret:
        raise RuntimeError("USER_HASH_SECRET is required")
    return secret.encode()


def user_key_from_telegram_id(telegram_id) -> str:
    """HMAC-SHA256 от telegram_id. Реальный PII в аналитике не хранится."""
    msg = str(telegram_id).encode()
    return hmac.new(_hash_secret(), msg, hashlib.sha256).hexdigest()[:32]


def new_tracking_token(prefix: str = "p") -> str:
    return f"{prefix}_{secrets.token_hex(3)}"


def build_deeplink(bot_username: str, token: str) -> str:
    return f"https://t.me/{bot_username}?start={token}"
